app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt

```diff
@@ -160,35 +165,47 @@ private val Event.wrapper
 
 class GleanMetricsService(private val context: Context) : MetricsService {
     private var initialized = false
+    /*
+     * We need to keep an eye on when we are done starting so that we don't
+     * accidentally stop ourselves before we've ever started.
+     */
+    private lateinit var starter: Job
 
     override fun start() {
-        Glean.setUploadEnabled(true)
+
         if (initialized) return
+        initialized = true
 
-        Glean.initialize(context)
+        starter = CoroutineScope(Dispatchers.Default).launch {
+            Glean.initialize(context)
 
-        Metrics.apply {
-            defaultBrowser.set(Browsers.all(context).isDefaultBrowser)
-            defaultMozBrowser.set(MozillaProductDetector.getMozillaBrowserDefault(context) ?: "")
-            mozillaProducts.set(MozillaProductDetector.getInstalledMozillaProducts(context))
-        }
+            Metrics.apply {
+                defaultBrowser.set(Browsers.all(context).isDefaultBrowser)
+                defaultMozBrowser.set(MozillaProductDetector.getMozillaBrowserDefault(context) ?: "")
+                mozillaProducts.set(MozillaProductDetector.getInstalledMozillaProducts(context))
+            }
 
-        SearchDefaultEngine.apply {
-            val defaultEngine = context
-                .components
-                .search
-                .searchEngineManager
-                .defaultSearchEngine ?: return@apply
+            SearchDefaultEngine.apply {
+                val defaultEngine = context
+                    .components
+                    .search
+                    .searchEngineManager
+                    .defaultSearchEngine ?: return@apply
 
-            code.set(defaultEngine.identifier)
-            name.set(defaultEngine.name)
-            submissionUrl.set(defaultEngine.buildSearchUrl(""))
-        }
+                code.set(defaultEngine.identifier)
+                name.set(defaultEngine.name)
+                submissionUrl.set(defaultEngine.buildSearchUrl(""))
 
-        initialized = true
+                Glean.setUploadEnabled(true)
+            }
+        }
     }
 
     override fun stop() {
+        /*
+         * We cannot stop until we're done starting.
+         */
+        runBlocking { starter.join(); }
```

Comment:
Aside: semi-colons are rarely used to end lines in Kotlin. They are only used when you want to put two statements on the same line, but are considered poor style inherited from Java.